lib/src/rules/use_iterables.dart

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:analyzer/dart/element/type_provider.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Use lazy iterables.';
+
+const _details = r'''
+**DO** use lazy iterables created by `iterable.map`. Since they are lazily
+instantiated, the `map` function will not be evaluated until use which can be
+confusing.  Alternatively, consider preferring `forEach`.
+
+**BAD:**
+```dart
+[1, 2, 3].map(print);
+```
+
+**GOOD:**
+```dart
+[1, 2, 3].forEach(print);
+[1, 2, 3].map(print).last;
+[1, 2, 3].map(print).toList();
```

Comment:
I'm not sure I can agree that the last two are examples of "good" code. :-)

But consider also proposing a simple `for` loop, which is equivalent to `forEach` when a closure is passed in rather than a tear-off.

Reply:
Yeah, I really dislike the last two but they are common work-arounds.  I'll defer to @natebosch, @jakemac53, @lrhn et al for whether we should exclude the work-arounds, common as they are.  (Alternatively, maybe some descriptive text?)

Reply:
I agree that we should not include those work-arounds in the "good" section.

I would use

```
[1, 2, 3].forEach(print);
for (var element in [1, 2, 3]) {
  print(element);
}
```

Reply:
Done.  Thanks!

Reply:
Agree with Nate. 
Using `.toList()` is wasteful, but otherwise harmless.
Using `.last` is plain wrong. If the receiver of the `map` call was a list, then `.last` will only call the map function on the last element of the list: `[1, 2, 3].map(print).last` only prints `3`.